File: ts/src/extensions/hooks.ts
import type { RuntimeComponentScope } from "../runtimes/component-lifecycle.js";
import type { RuntimeCompositionInventory } from "../runtimes/composition-observability.js";

export enum HookEvents {
  RUN_START = "run_start",
  RUN_END = "run_end",
  GENERATION_START = "generation_start",
  GENERATION_END = "generation_end",
  CONTEXT_COMPONENTS = "context_components",
  CONTEXT = "context",
  BEFORE_COMPACTION = "before_compaction",
  AFTER_COMPACTION = "after_compaction",
  BEFORE_PROVIDER_REQUEST = "before_provider_request",
  AFTER_PROVIDER_RESPONSE = "after_provider_response",
  BEFORE_JUDGE = "before_judge",
  AFTER_JUDGE = "after_judge",
  ARTIFACT_WRITE = "artifact_write",
}

export interface HookResultOptions {
  payload?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
  replacePayload?: boolean;
  block?: boolean;
  reason?: string;
}

export class HookResult {
  readonly payload: Record<string, unknown> | null;
  readonly metadata: Record<string, unknown> | null;
  readonly replacePayload: boolean;
  readonly block: boolean;
  readonly reason: string;

  constructor(opts: HookResultOptions = {}) {
    this.payload = opts.payload ?? null;
    this.metadata = opts.metadata ?? null;
    this.replacePayload = opts.replacePayload ?? false;
    this.block = opts.block ?? false;
    this.reason = opts.reason ?? "";
  }
}

export interface HookError {
  eventName: string;
  handler: string;
  message: string;
}

export class HookEvent {
  readonly name: string;
  payload: Record<string, unknown>;
  metadata: Record<string, unknown>;
  errors: HookError[];
  blocked: boolean;
  blockReason: string;

  constructor(
    name: HookEvents | string,
    payload: Record<string, unknown> = {},
    metadata: Record<string, unknown> = {},
  ) {
    this.name = eventName(name);
    this.payload = { ...payload };
    this.metadata = { ...metadata };
    this.errors = [];
    this.blocked = false;
    this.blockReason = "";
  }

  raiseIfBlocked(): void {
    if (this.blocked) {
      throw eventBlockError(this);
    }
  }
}

export type HookHandler = (event: HookEvent) => HookResult | Record<string, unknown> | undefined | null;
export type HookDisposer = () => void | Promise<void>;

interface HookRegistration {
  readonly handler: HookHandler;
  active: boolean;
}

export function eventName(value: HookEvents | string): string {
  return typeof value === "string" ? value : String(value);
}

export function eventBlockError(event: HookEvent): Error {
  const reason = event.blockReason ? `: ${event.blockReason}` : "";
  return new Error(`extension hook blocked ${event.name}${reason}`);
}

export class HookBus {
  readonly failFast: boolean;
  readonly loadedExtensions: string[];
  private handlers: Map<string, HookRegistration[]>;

  constructor(opts: { failFast?: boolean; loadedExtensions?: string[] } = {}) {
    this.failFast = opts.failFast ?? false;
    this.loadedExtensions = [...(opts.loadedExtensions ?? [])];
    this.handlers = new Map();
  }

  on(name: HookEvents | string, handler: HookHandler): HookHandler {
    this.register(name, handler);
    return handler;
  }

  subscribe(name: HookEvents | string, handler: HookHandler): HookDisposer {
    return this.register(name, handler);
  }

  private register(name: HookEvents | string, handler: HookHandler): HookDisposer {
    const normalized = eventName(name);
    const handlers = this.handlers.get(normalized) ?? [];
    const registration: HookRegistration = { handler, active: true };
    handlers.push(registration);
    this.handlers.set(normalized, handlers);
    return () => {
      if (!registration.active) return;
      registration.active = false;
      const current = this.handlers.get(normalized);
      if (!current) return;
      const index = current.indexOf(registration);
      if (index !== -1) current.splice(index, 1);
      if (current.length === 0) this.handlers.delete(normalized);
    };
  }

  hasHandlers(name: HookEvents | string): boolean {
    const normalized = eventName(name);
    return Boolean(this.handlers.get(normalized)?.length || this.handlers.get("*")?.length);
  }

  emit(
    name: HookEvents | string,
    payload: Record<string, unknown> = {},
    opts: { metadata?: Record<string, unknown> } = {},
  ): HookEvent {
    const normalized = eventName(name);
    const event = new HookEvent(normalized, payload, opts.metadata ?? {});
    const handlers = [
      ...(this.handlers.get(normalized) ?? []),
      ...(this.handlers.get("*") ?? []),
    ];

    for (const registration of handlers) {
      if (!registration.active) continue;
      const handler = registration.handler;
      try {
        const result = handler(event);
        applyHookResult(event, result);
      } catch (error) {
        if (this.failFast) {
          throw error;
        }
        event.errors.push({
          eventName: normalized,
          handler: handlerName(handler),
          message: error instanceof Error ? error.message : String(error),
        });
      }
      if (event.blocked) {
        break;
      }
    }
    return event;
  }
}

export class ExtensionAPI {
  readonly bus: HookBus;
  readonly scope?: RuntimeComponentScope;
  readonly compositionInventory?: RuntimeCompositionInventory;

  constructor(
    bus: HookBus,
    scope?: RuntimeComponentScope,
    compositionInventory?: RuntimeCompositionInventory,
  ) {
    this.bus = bus;
    this.scope = scope;
    this.compositionInventory = compositionInventory;
  }

  on(name: HookEvents | string, handler: HookHandler): HookHandler;
  on(name: HookEvents | string): (handler: HookHandler) => HookHandler;
  on(
    name: HookEvents | string,
    handler?: HookHandler,
  ): HookHandler | ((handler: HookHandler) => HookHandler) {
    if (handler) {
      this.register(name, handler);
      return handler;
    }
    return (actual: HookHandler) => {
      this.register(name, actual);
      return actual;
    };
  }

  emit(
    name: HookEvents | string,
    payload: Record<string, unknown> = {},
    opts: { metadata?: Record<string, unknown> } = {},
  ): HookEvent {
    return this.bus.emit(name, payload, opts);
  }

  private register(name: HookEvents | string, handler: HookHandler): void {
    const unsubscribe = this.bus.subscribe(name, handler);
    if (this.scope) {
      try {
        this.scope.defer(unsubscribe);
        this.compositionInventory?.own(this.scope, {
          kind: "hook",
          resourceId: eventName(name),
          effectClass: "reversible",
        });
      } catch (error) {
        void unsubscribe();
        throw error;
      }
    }
  }
}

function applyHookResult(
  event: HookEvent,
  result: HookResult | Record<string, unknown> | undefined | null,
): void {
  if (result === undefined || result === null) {
    return;
  }
  if (result instanceof HookResult) {
    if (result.payload) {
      if (result.replacePayload) {
        event.payload = { ...result.payload };
      } else {
        event.payload = { ...event.payload, ...result.payload };
      }
    }
    if (result.metadata) {
      event.metadata = { ...event.metadata, ...result.metadata };
    }
    if (result.block) {
      event.blocked = true;
      event.blockReason = result.reason;
    }
    return;
  }
  event.payload = { ...event.payload, ...result };
}

function handlerName(handler: HookHandler): string {
  return handler.name || "anonymous_hook_handler";
}

File: autocontext/src/autocontext/extensions/hooks.py
from __future__ import annotations

from collections.abc import Callable, Iterator, Mapping
from contextlib import contextmanager
from contextvars import ContextVar
from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any


class HookEvents(StrEnum):
    """Stable built-in hook names inspired by Pi's extension event contracts."""

    RUN_START = "run_start"
    RUN_END = "run_end"
    GENERATION_START = "generation_start"
    GENERATION_END = "generation_end"
    CONTEXT_COMPONENTS = "context_components"
    CONTEXT = "context"
    BEFORE_COMPACTION = "before_compaction"
    AFTER_COMPACTION = "after_compaction"
    BEFORE_PROVIDER_REQUEST = "before_provider_request"
    AFTER_PROVIDER_RESPONSE = "after_provider_response"
    BEFORE_JUDGE = "before_judge"
    AFTER_JUDGE = "after_judge"
    ARTIFACT_WRITE = "artifact_write"


@dataclass(frozen=True, slots=True)
class HookResult:
    """A hook's requested changes to the in-flight event.

    ``payload`` is merged into the event payload by default. Set
    ``replace_payload`` when a hook wants to replace the complete payload.
    """

    payload: Mapping[str, Any] | None = None
    metadata: Mapping[str, Any] | None = None
    replace_payload: bool = False
    block: bool = False
    reason: str = ""


@dataclass(frozen=True, slots=True)
class HookError:
    """Non-fatal hook failure captured when the bus is not fail-fast."""

    event_name: str
    handler: str
    message: str


@dataclass(slots=True)
class HookEvent:
    """Mutable event object passed to extension handlers."""

    name: str
    payload: dict[str, Any] = field(default_factory=dict)
    metadata: dict[str, Any] = field(default_factory=dict)
    errors: list[HookError] = field(default_factory=list)
    blocked: bool = False
    block_reason: str = ""

    def raise_if_blocked(self) -> None:
        if self.blocked:
            raise event_block_error(self)


HookHandler = Callable[[HookEvent], HookResult | Mapping[str, Any] | None]


def event_name(value: HookEvents | str) -> str:
    return value.value if isinstance(value, HookEvents) else str(value)


def event_block_error(event: HookEvent) -> RuntimeError:
    reason = f": {event.block_reason}" if event.block_reason else ""
    return RuntimeError(f"extension hook blocked {event.name}{reason}")


def _handler_name(handler: HookHandler) -> str:
    module = getattr(handler, "__module__", "")
    qualname = getattr(handler, "__qualname__", "")
    if module and qualname:
        return f"{module}.{qualname}"
    return repr(handler)


class HookBus:
    """Ordered, fail-open extension hook bus.

    Hooks run in registration order. They may mutate the event directly or return
    a ``HookResult`` / mapping. Handler exceptions are recorded on the event by
    default so production runs do not die because an optional extension failed.
    """

    def __init__(self, *, fail_fast: bool = False) -> None:
        self.fail_fast = fail_fast
        self._handlers: dict[str, list[HookHandler]] = {}

    def on(self, name: HookEvents | str, handler: HookHandler) -> HookHandler:
        self._handlers.setdefault(event_name(name), []).append(handler)
        return handler

    def has_handlers(self, name: HookEvents | str) -> bool:
        normalized = event_name(name)
        return bool(self._handlers.get(normalized) or self._handlers.get("*"))

    def emit(
        self,
        name: HookEvents | str,
        payload: Mapping[str, Any] | None = None,
        *,
        metadata: Mapping[str, Any] | None = None,
    ) -> HookEvent:
        normalized = event_name(name)
        event = HookEvent(
            name=normalized,
            payload=dict(payload or {}),
            metadata=dict(metadata or {}),
        )
        handlers = [*self._handlers.get(normalized, ()), *self._handlers.get("*", ())]
        for handler in handlers:
            try:
                result = handler(event)
            except Exception as exc:
                if self.fail_fast:
                    raise
                event.errors.append(
                    HookError(
                        event_name=normalized,
                        handler=_handler_name(handler),
                        message=str(exc),
                    )
                )
                continue
            self._apply_result(event, result)
            if event.blocked:
                break
        return event

    @staticmethod
    def _apply_result(event: HookEvent, result: HookResult | Mapping[str, Any] | None) -> None:
        if result is None:
            return
        if isinstance(result, HookResult):
            if result.payload is not None:
                if result.replace_payload:
                    event.payload = dict(result.payload)
                else:
                    event.payload.update(result.payload)
            if result.metadata is not None:
                event.metadata.update(result.metadata)
            if result.block:
                event.blocked = True
                event.block_reason = result.reason
            return
        event.payload.update(dict(result))


_CURRENT_HOOK_BUS: ContextVar[HookBus | None] = ContextVar("autocontext_current_hook_bus", default=None)


def get_current_hook_bus() -> HookBus | None:
    return _CURRENT_HOOK_BUS.get()


@contextmanager
def active_hook_bus(hook_bus: HookBus | None) -> Iterator[None]:
    if hook_bus is None:
        yield
        return
    token = _CURRENT_HOOK_BUS.set(hook_bus)
    try:
        yield
    finally:
        _CURRENT_HOOK_BUS.reset(token)


class ExtensionAPI:
    """Small registration facade passed to extension modules."""

    def __init__(self, bus: HookBus) -> None:
        self.bus = bus

    def on(
        self,
        name: HookEvents | str,
        handler: HookHandler | None = None,
    ) -> HookHandler | Callable[[HookHandler], HookHandler]:
        if handler is not None:
            return self.bus.on(name, handler)

        def decorator(actual: HookHandler) -> HookHandler:
            return self.bus.on(name, actual)

        return decorator

    def emit(
        self,
        name: HookEvents | str,
        payload: Mapping[str, Any] | None = None,
        *,
        metadata: Mapping[str, Any] | None = None,
    ) -> HookEvent:
        return self.bus.emit(name, payload, metadata=metadata)
